subprojects/core/src/main/java/org/gradle/api/internal/attributes/DefaultMutableAttributeContainer.java

```diff
@@ -58,6 +60,15 @@ public String toString() {
         return this;
     }
 
+    @Override
+    public AttributeContainer attribute(String name, String value) {
+        Attribute<String> key = Attribute.of(name, String.class);
+        assertAttributeConstraints(value, key);
+        checkInsertionAllowed(key);
+        state = cache.concat(state, key, new CoercingStringValueSnapshot(value, NamedObjectInstantiator.INSTANCE));
```

Comment:
:x: What if you want to use a `String` attribute? It would create a coercing string value, when all you need is a plain `String`.

Reply:
I think this is fine. It's the same situation as if we read a `String` attribute from metadata. The moment we read it, we also do not know if it is a `String` or another type that needs coercing.

I added a test case for this.